src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -137,15 +137,32 @@ double LMNNFunction<MetricType>::Evaluate(const arma::mat& transformation)
         bool exactEval = true;
 
         // Bounds for eval.
-        if (transformationOld.n_elem != 0 && !std::isnan(evalOld[i](j, l)))
+        if (transformationOld.n_elem != 0 && !std::isnan(evalOld(j, l, i)))
         {
-          eval = evalOld[i](j, l) + transformationDiff *
-            (norm(targetNeighbors(j, i)) + norm(impostors(l, i)) +
+          double maxNorm;
+          if (!std::isnan(maxImpNorm(l, i)))
+            maxNorm = std::max(norm(impostors(l, i)), maxImpNorm(l, i));
+          else
+            maxNorm = norm(impostors(l, i));
+
+          // Update cache max impostor norm.
+          maxImpNorm(l, i) = maxNorm;
+
+          eval = evalOld(j, l, i) + transformationDiff *
```

Comment:
This is a minor point, but since Armadillo is column-major, points in the same column end up being contiguous in memory.  Since we're looping over `i`, then `j`, then `l`, the best access pattern for `evalOld` will be if we are accessing as `evalOld(l, j, i)` (i.e. `l` corresponds to the row, `j` to the column, `i` to the slice).